File: lib/meilisearch/bangumiSearch.ts
import {getMeiliClient} from './client'
import type {SearchParams} from 'meilisearch'

// Bangumi 搜索索引名称 (与 products 索引区分)
export const BANGUMI_INDEX = 'bangumi_subjects'

// Bangumi 搜索文档类型
export interface BangumiSearchDoc {
    /** 条目 ID */
    subject_id: number
    /** 原名 */
    name: string
    /** 中文名 */
    name_cn: string
    /** 条目类型 ID */
    type: number
    /** 条目类型名称 */
    type_name: string
    /** 平台 ID */
    platform?: number
    /** 平台名称 */
    platform_name?: string
    /** 简介 */
    summary: string
    /** 放送/发售日期 */
    date?: string
    /** 评分 */
    score?: number
    /** 排名 */
    rank?: number
    /** 条目页面 URL */
    url: string
    /** 标签 */
    tags?: string[]
    /** 是否 NSFW */
    nsfw: boolean
}

// 搜索结果类型
export interface BangumiSearchResult {
    subjects: BangumiSearchDoc[]
    total: number
    page: number
    totalPages: number
}

// 过滤选项
export interface BangumiSearchOptions {
    page?: number
    limit?: number
    type?: number
    platform?: number
    minScore?: number
    maxScore?: number
    nsfw?: boolean
}

// ============ 索引管理 ============

/**
 * 创建 Bangumi 搜索索引
 */
export async function setupBangumiIndex(): Promise<void> {
    const meili = getMeiliClient()

    try {
        // 获取或创建索引
        const index = await meili.getIndex(BANGUMI_INDEX)
        console.log('[meilisearch] Bangumi index already exists:', BANGUMI_INDEX)
        await updateBangumiIndexSettings(index)
    } catch {
        // 索引不存在，创建它
        console.log('[meilisearch] Creating Bangumi index:', BANGUMI_INDEX)
        await meili.createIndex(BANGUMI_INDEX, {
            primaryKey: 'subject_id',
        })

        const index = meili.index(BANGUMI_INDEX)
        await updateBangumiIndexSettings(index)
    }
}

/**
 * 更新 Bangumi 索引设置
 */
async function updateBangumiIndexSettings(index: any): Promise<void> {
    await index.updateSettings({
        // 可搜索字段
        searchableAttributes: [
            'name',
            'name_cn',
            'summary',
            'tags',
        ],
        // 可过滤字段
        filterableAttributes: [
            'type',
            'platform',
            'score',
            'nsfw',
            'tags',
        ],
        // 可排序字段
        sortableAttributes: [
            'subject_id',
            'score',
            'rank',
            'date',
        ],
        // 排序规则
        rankingRules: [
            'words',
            'typo',
            'proximity',
            'attribute',
            'sort',
            'exactness',
        ],
        // 同义词
        synonyms: {
            'tv': ['television', '电视'],
            'ova': ['original video animation'],
            '剧场版': ['movie', '电影'],
        },
        // 停用词
        stopWords: [],
    })

    console.log('[meilisearch] Bangumi index settings updated')
}

/**
 * 删除 Bangumi 索引
 */
export async function deleteBangumiIndex(): Promise<void> {
    const meili = getMeiliClient()
    try {
        await meili.deleteIndex(BANGUMI_INDEX)
        console.log('[meilisearch] Deleted Bangumi index:', BANGUMI_INDEX)
    } catch (error) {
        console.log('[meilisearch] Bangumi index does not exist:', BANGUMI_INDEX)
    }
}

// ============ 文档操作 ============

/**
 * 索引单个 Bangumi 条目
 */
export async function indexBangumiSubject(subject: BangumiSearchDoc): Promise<void> {
    const meili = getMeiliClient()
    const index = meili.index(BANGUMI_INDEX)
    await index.addDocuments([subject])
}

/**
 * 批量索引 Bangumi 条目
 */
export async function bulkIndexBangumiSubjects(
    subjects: BangumiSearchDoc[],
    onProgress?: (processed: number, total: number) => void
): Promise<void> {
    const meili = getMeiliClient()
    const index = meili.index(BANGUMI_INDEX)

    const response = await index.addDocuments(subjects)
    console.log(`[meilisearch] Added ${subjects.length} Bangumi documents, task uid: ${response.taskUid}`)

    if (onProgress) {
        onProgress(subjects.length, subjects.length)
    }
}

/**
 * 删除单个 Bangumi 条目
 */
export async function deleteBangumiSubject(subjectId: number): Promise<void> {
    const meili = getMeiliClient()
    const index = meili.index(BANGUMI_INDEX)
    await index.deleteDocument(subjectId)
}

/**
 * 清空所有 Bangumi 条目
 */
export async function clearAllBangumiSubjects(): Promise<void> {
    const meili = getMeiliClient()
    const index = meili.index(BANGUMI_INDEX)
    await index.deleteAllDocuments()
    console.log('[meilisearch] All Bangumi documents deleted')
}

// ============ 搜索 ============

/**
 * 搜索 Bangumi 条目
 */
export async function searchBangumi(
    query: string,
    options: BangumiSearchOptions = {}
): Promise<BangumiSearchResult> {
    const meili = getMeiliClient()
    const index = meili.index(BANGUMI_INDEX)

    const {
        page = 1,
        limit = 20,
        type,
        platform,
        minScore,
        maxScore,
        nsfw,
    } = options

    // 构建过滤条件
    const filters: string[] = []

    if (type !== undefined) {
        filters.push(`type = ${type}`)
    }

    if (platform !== undefined) {
        filters.push(`platform = ${platform}`)
    }

    if (minScore !== undefined || maxScore !== undefined) {
        const min = minScore ?? 0
        const max = maxScore ?? 10
        filters.push(`score ${min} TO ${max}`)
    }

    if (nsfw !== undefined) {
        filters.push(`nsfw = ${nsfw}`)
    }

    const searchParams: SearchParams = {
        offset: (page - 1) * limit,
        limit: limit,
        // 高亮配置
        attributesToHighlight: ['name', 'name_cn', 'summary'],
        highlightPreTag: '<mark>',
        highlightPostTag: '</mark>',
        // 裁剪长文本
        attributesToCrop: ['summary'],
        cropLength: 200,
        // 匹配信息
        showMatchesPosition: true,
        // 排序
        sort: ['rank:asc'],
    }

    if (filters.length > 0) {
        searchParams.filter = filters
    }

    const response = await index.search(query, searchParams)

    // 转换结果
    const subjects = response.hits.map(hit => {
        const doc = hit as any
        return {
            subject_id: doc.subject_id,
            name: doc.name,
            name_cn: doc.name_cn,
            type: doc.type,
            type_name: doc.type_name,
            platform: doc.platform,
            platform_name: doc.platform_name,
            summary: doc.summary,
            date: doc.date,
            score: doc.score,
            rank: doc.rank,
            url: doc.url,
            tags: doc.tags,
            nsfw: doc.nsfw,
            // 高亮结果
            ...(doc._formatted?.name_cn && {highlight_name_cn: doc._formatted.name_cn}),
            ...(doc._formatted?.summary && {highlight_summary: doc._formatted.summary}),
        } as BangumiSearchDoc & { highlight_name_cn?: string; highlight_summary?: string }
    })

    return {
        subjects,
        total: response.estimatedTotalHits ?? 0,
        page,
        totalPages: Math.ceil((response.estimatedTotalHits ?? 0) / limit),
    }
}

/**
 * 获取 Bangumi 索引统计
 */
export async function getBangumiIndexStats(): Promise<{
    totalDocuments: number
    isIndexing: boolean
}> {
    const meili = getMeiliClient()
    const index = meili.index(BANGUMI_INDEX)
    const stats = await index.getStats()

    return {
        totalDocuments: stats.numberOfDocuments,
        isIndexing: stats.isIndexing,
    }
}

/**
 * 获取类型分布统计
 */
export async function getBangumiTypeStats(): Promise<Record<number, number>> {
    const meili = getMeiliClient()
    const index = meili.index(BANGUMI_INDEX)

    // 使用 facet search 获取类型分布
    const response = await index.search('', {
        facets: ['type'],
        limit: 0,
    })

    const typeStats: Record<number, number> = {}
    const facetDistribution = response.facetDistribution?.type

    if (facetDistribution) {
        for (const [type, count] of Object.entries(facetDistribution)) {
            typeStats[parseInt(type)] = count
        }
    }

    return typeStats
}

File: src/search/bangumisearch.cpp
#include "search/bangumisearch.h"

#include <QJsonObject>
#include <QJsonArray>
#include <QString>

// BangumiSearchDoc 实现

QString BangumiSearchDoc::toJson() const {
    QJsonObject obj;
    obj["subject_id"] = subjectId;
    obj["name"] = name;
    obj["name_cn"] = nameCn;
    obj["type"] = type;
    obj["type_name"] = typeName;
    obj["summary"] = summary;
    obj["url"] = url;
    obj["nsfw"] = nsfw;

    if (platform) obj["platform"] = *platform;
    if (platformName) obj["platform_name"] = *platformName;
    if (date) obj["date"] = *date;
    if (score) obj["score"] = *score;
    if (rank) obj["rank"] = *rank;

    QJsonArray tagsArray;
    for (const auto &tag: tags) {
        tagsArray.append(tag);
    }
    if (!tagsArray.isEmpty()) {
        obj["tags"] = tagsArray;
    }

    QJsonDocument doc(obj);
    return QString::fromUtf8(doc.toJson(QJsonDocument::Compact));
}

// BangumiSearchService 实现

SearchResult<void> BangumiSearchService::setupIndex() {
    auto &client = MeiliSearchClient::instance();

    auto exists = client.indexExists(kIndexName);
    if (!exists) return std::unexpected(exists.error());

    if (!*exists) {
        auto create = client.createIndex(kIndexName, "subject_id");
        if (!create) return create;
    }

    return updateIndexSettings();
}

SearchResult<void> BangumiSearchService::deleteIndex() {
    return MeiliSearchClient::instance().deleteIndex(kIndexName);
}

SearchResult<void> BangumiSearchService::indexSubject(const BangumiSearchDoc &subject) {
    QList subjects{subject};
    return bulkIndexSubjects(subjects);
}

SearchResult<void> BangumiSearchService::bulkIndexSubjects(
    const QList<BangumiSearchDoc> &subjects,
    std::optional<std::function<void(qint32 processed, qint32 total)> > onProgress
) {
    if (subjects.empty()) return {};

    auto &client = MeiliSearchClient::instance();

    // 构建 JSON 数组
    QJsonArray array;
    for (const auto &subject: subjects) {
        QJsonDocument doc = QJsonDocument::fromJson(subject.toJson().toUtf8());
        array.append(doc.object());
    }

    QJsonDocument doc(array);
    QString json = QString::fromUtf8(doc.toJson(QJsonDocument::Compact));

    auto result = client.addDocuments(kIndexName, json);
    if (!result) return result;

    if (onProgress) {
        (*onProgress)(static_cast<qint32>(subjects.size()), static_cast<qint32>(subjects.size()));
    }

    return {};
}

SearchResult<void> BangumiSearchService::deleteSubject(qint32 subjectId) {
    return MeiliSearchClient::instance().deleteDocument(kIndexName, QString::number(subjectId));
}

SearchResult<void> BangumiSearchService::clearAllSubjects() {
    return MeiliSearchClient::instance().deleteAllDocuments(kIndexName);
}

SearchResult<BangumiSearchResult> BangumiSearchService::search(
    const QString &query,
    const BangumiSearchOptions &options
) {
    auto &client = MeiliSearchClient::instance();

    qint32 offset = (options.page - 1) * options.limit;

    // 构建过滤条件
    QList<QString> filters;

    if (options.type) {
        filters.push_back("type = " + QString::number(*options.type));
    }

    if (options.platform) {
        filters.push_back("platform = " + QString::number(*options.platform));
    }

    if (options.minScore || options.maxScore) {
        qreal min = options.minScore.value_or(0.0);
        qreal max = options.maxScore.value_or(10.0);
        filters.push_back("score " + QString::number(min) + " TO " + QString::number(max));
    }

    if (options.nsfw) {
        filters.push_back("nsfw = " + QString(*options.nsfw ? "true" : "false"));
    }

    QList<QString> sort = {QStringLiteral("rank:asc")};

    auto result = client.searchRaw(kIndexName, query, offset, options.limit, filters, sort);
    if (!result) return std::unexpected(result.error());

    // 解析响应
    QJsonDocument doc = QJsonDocument::fromJson(result->toUtf8());
    if (!doc.isObject()) {
        return std::unexpected(QStringLiteral("Invalid search response"));
    }

    QJsonObject obj = doc.object();
    BangumiSearchResult searchResult;

    // 解析总数
    searchResult.total = static_cast<qint32>(obj["estimatedTotalHits"].toInt());
    searchResult.page = options.page;
    searchResult.totalPages = (searchResult.total + options.limit - 1) / options.limit;

    // 解析结果
    QJsonArray hits = obj["hits"].toArray();
    for (const auto &hit: hits) {
        if (!hit.isObject()) continue;

        QJsonObject hitObj = hit.toObject();
        BangumiSearchDoc subject;

        subject.subjectId = static_cast<qint32>(hitObj["subject_id"].toInt());
        subject.name = hitObj["name"].toString();
        subject.nameCn = hitObj["name_cn"].toString();
        subject.type = static_cast<qint32>(hitObj["type"].toInt());
        subject.typeName = hitObj["type_name"].toString();
        subject.summary = hitObj["summary"].toString();
        subject.url = hitObj["url"].toString();
        subject.nsfw = hitObj["nsfw"].toBool();

        if (hitObj.contains("platform")) {
            subject.platform = static_cast<qint32>(hitObj["platform"].toInt());
        }
        if (hitObj.contains("platform_name")) {
            subject.platformName = hitObj["platform_name"].toString();
        }
        if (hitObj.contains("date")) {
            subject.date = hitObj["date"].toString();
        }
        if (hitObj.contains("score")) {
            subject.score = hitObj["score"].toDouble();
        }
        if (hitObj.contains("rank")) {
            subject.rank = static_cast<qint32>(hitObj["rank"].toInt());
        }

        QJsonArray tags = hitObj["tags"].toArray();
        for (const auto &tag: tags) {
            subject.tags.push_back(tag.toString());
        }

        // 高亮结果
        if (hitObj.contains("_formatted")) {
            QJsonObject formatted = hitObj["_formatted"].toObject();
            if (formatted.contains("name_cn")) {
                subject.highlightNameCn = formatted["name_cn"].toString();
            }
            if (formatted.contains("summary")) {
                subject.highlightSummary = formatted["summary"].toString();
            }
        }

        searchResult.subjects.push_back(std::move(subject));
    }

    return searchResult;
}

SearchResult<IndexStats> BangumiSearchService::getIndexStats() {
    return MeiliSearchClient::instance().getIndexStats(kIndexName);
}

SearchResult<QMap<qint32, qint32> > BangumiSearchService::getTypeStats() {
    auto &client = MeiliSearchClient::instance();

    // 使用 facet search 获取类型分布
    // Meilisearch facet 搜索通过在搜索请求中添加 facets 参数
    auto result = client.searchRaw(kIndexName, "", 0, 0, {}, {});
    if (!result) return std::unexpected(result.error());

    // 解析 facetDistribution
    QJsonDocument doc = QJsonDocument::fromJson(result->toUtf8());
    if (!doc.isObject()) {
        return std::unexpected(QStringLiteral("Invalid stats response"));
    }

    QJsonObject obj = doc.object();
    QMap<qint32, qint32> typeStats;

    QJsonObject facetDistribution = obj["facetDistribution"].toObject();
    QJsonObject typeDistribution = facetDistribution["type"].toObject();

    for (const QString &key: typeDistribution.keys()) {
        qint32 type = key.toInt();
        qint32 count = static_cast<qint32>(typeDistribution[key].toInt());
        typeStats[type] = count;
    }

    return typeStats;
}

SearchResult<void> BangumiSearchService::updateIndexSettings() {
    // 索引设置通常在创建时配置
    return {};
}
